storage/index.go

```diff
@@ -265,10 +265,15 @@ func (i *nsIndex) Query(
 		return index.QueryResults{}, err
 	}
 
+	idxIter := index.NewIterator(i.nsID, iter, i.opts)
+
+	ctx.RegisterFinalizer(resource.FinalizerFn(func() {
+		idxIter.Finalize()
+		exec.Close()
+	}))
```

Comment:
nit: To avoid a lambda alloc (you'll still need to do a method bind alloc but it's slightly cheaper I think) you could probably:
```go
ctx.RegisterFinalizer(idxIter)
ctx.RegisterFinalizer(resource.FinalizerFn(exec.Close))
```

Reply:
sure

Reply:
Actually can't do it. `exec.Close()` returns an error so i need the lambda alloc unfortunately.

Reply:
Discussed, exec.Close returns an error.  We can just keep an alloc for now.